### PR TITLE
Remove WebSocketComponents & HouseKeeper on Server restart.

### PR DESCRIPTION
**Pull request by lachlan-roberts**
If I start and stop jetty in a loop 100 times I get a build up of `WebSocketServerComponents` and `HouseKeeper`.

This PR ensures that `WebSocketServerComponents` does not add a durable `ServletContextListener`.
And that `HouseKeeper` beans are removed from the `DefaultSessionIdManager` in `doStop()`.

```
+= o.e.j.s.ServletContextHandler@3ba987b8{/,null,AVAILABLE} - STARTED
|  += org.eclipse.jetty.server.session.SessionHandler1902237905==dftMaxIdleSec=-1 - STARTED
|  |  += ServletHandler@50029372{STARTED} - STARTED
|  |  |  +> listeners ServletHandler@50029372{STARTED} size=10
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@e3b3b2f{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@50f6ac94{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@6cc4cdb9{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@28194a50{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@7f2cfe3f{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@1a5b6f42{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@5038d0b5{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@32115b28{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@2ad48653{src=JAVAX_API:null} - STARTED
|  |  |  |  +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@6bb4dd34{src=JAVAX_API:null} - STARTED
|  |  |  +> filters ServletHandler@50029372{STARTED} size=1
|  |  |  |  +> org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter==org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter@1b11171f{inst=true,async=true,src=EMBEDDED:null} - STARTED
|  |  |  |     +> org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter@7d9f158f
|  |  |  |        +> org.eclipse.jetty.websocket.core.server.WebSocketMappings@45efd90f
|  |  |  |           +> PathMappings[size=1]
|  |  |  |              +> java.util.TreeSet@6d(size=1)
|  |  |  |                 +: MappedResource[pathSpec=ServletPathSpec@4e{/},resource=CreatorNegotiator@383dc82c{org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer$$Lambda$353/0x00000008001ea840@4a07d605,JettyServerFrameHandlerFactory@74287ea3{STARTED}}]
|  |  |  +> filterMappings ServletHandler@50029372{STARTED} size=1
|  |  |  |  +> [/*]/[]/[REQUEST]=>org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter
|  |  |  +> servlets ServletHandler@50029372{STARTED} size=1
|  |  |  |  +> org.eclipse.jetty.servlet.ServletHandler$Default404Servlet-14bdbc74==org.eclipse.jetty.servlet.ServletHandler$Default404Servlet@aaf08b97{jsp=null,order=-1,inst=false,async=true,src=EMBEDDED:null} - STARTED
|  |  |  |     +> class org.eclipse.jetty.servlet.ServletHandler$Default404Servlet
|  |  |  +> servletMappings ServletHandler@50029372{STARTED} size=1
|  |  |  |  +> [/]=>org.eclipse.jetty.servlet.ServletHandler$Default404Servlet-14bdbc74
|  |  |  +> durable ServletHandler@50029372{STARTED} size=11
|  |  |     +> org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter==org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter@1b11171f{inst=true,async=true,src=EMBEDDED:null} - STARTED
|  |  |     |  +> org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter@7d9f158f
|  |  |     |     +> org.eclipse.jetty.websocket.core.server.WebSocketMappings@45efd90f
|  |  |     |        +> PathMappings[size=1]
|  |  |     |           +> java.util.TreeSet@6d(size=1)
|  |  |     |              +: MappedResource[pathSpec=ServletPathSpec@4e{/},resource=CreatorNegotiator@383dc82c{org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer$$Lambda$353/0x00000008001ea840@4a07d605,JettyServerFrameHandlerFactory@74287ea3{STARTED}}]
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@e3b3b2f{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@50f6ac94{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@6cc4cdb9{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@28194a50{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@7f2cfe3f{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@1a5b6f42{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@5038d0b5{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@32115b28{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@2ad48653{src=JAVAX_API:null} - STARTED
|  |  |     +> org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@6bb4dd34{src=JAVAX_API:null} - STARTED
|  |  += org.eclipse.jetty.server.session.DefaultSessionCache@81d9a72[evict=-1,removeUnloadable=false,saveOnCreate=false,saveOnInactiveEvict=false] - STARTED
|  |  |  += org.eclipse.jetty.server.session.NullSessionDataStore@747f281[passivating=false,graceSec=3600] - STARTED
|  |  +~ DefaultSessionIdManager@1169afe1{STARTED}[worker=node0] - STARTED
|  += ServletContainerInitializerStarter@2ca26d77{STARTED} - STARTED
|  |  += ContainerInitializer{org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer,interested=[],applicable=[],annotated=[]} - STARTED
|  += JettyWebSocketServerContainer@70e38ce1{STARTED} - STARTED
|  |  += JettyServerFrameHandlerFactory@74287ea3{STARTED} - STARTED
|  |  |  +> java.util.concurrent.ConcurrentHashMap@0{size=0}
|  |  += SessionTracker@2ca923bb{STARTED} - STARTED
|  |     +> java.util.Collections$SetFromMap@0(size=0)
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@13df2a8c
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@1ebea008
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@72d6b3ba
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@1787f2a0
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@7de62196
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@163370c2
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@51bf5add
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@7905a0b8
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@35a3d49f
|  +- org.eclipse.jetty.websocket.core.server.WebSocketServerComponents$1@389b0789
|  +> No ClassLoader
|  +> handler attributes o.e.j.s.ServletContextHandler@3ba987b8{/,null,AVAILABLE} size=1
|  |  +> org.eclipse.jetty.server.Executor=QueuedThreadPool[qtp606508809]@24269709{STARTED,8<=15<=200,i=0,r=-1,q=0}[ReservedThreadExecutor@798162bc{s=0/20,p=0}]
|  +> context attributes o.e.j.s.ServletContextHandler@3ba987b8{/,null,AVAILABLE} size=4
|  |  +> org.eclipse.jetty.util.DecoratedObjectFactory=org.eclipse.jetty.util.DecoratedObjectFactory[decorators=0]
|  |  +> org.eclipse.jetty.websocket.api.WebSocketContainer=JettyWebSocketServerContainer@70e38ce1{STARTED}
|  |  +> org.eclipse.jetty.websocket.core.WebSocketComponents=WebSocketServerComponents@478db956{STARTED}
|  |  +> org.eclipse.jetty.websocket.core.server.WebSocketMappings=org.eclipse.jetty.websocket.core.server.WebSocketMappings@45efd90f
|  +> initparams o.e.j.s.ServletContextHandler@3ba987b8{/,null,AVAILABLE} size=0
+= ErrorHandler@6ca18a14{STARTED} - STARTED
+= InflaterPool@c667f46{STARTED,size=0,capacity=200} - STARTED
+= DeflaterPool@51bd8b5c{STARTED,size=0,capacity=200} - STARTED
+= DefaultSessionIdManager@1169afe1{STARTED}[worker=node0] - STARTED
|  += HouseKeeper@6e6f2380{STOPPED}[interval=600000, ownscheduler=false] - STOPPED
|  += HouseKeeper@71075444{STOPPED}[interval=660000, ownscheduler=false] - STOPPED
|  += HouseKeeper@4e5ed836{STOPPED}[interval=660000, ownscheduler=false] - STOPPED
|  += HouseKeeper@48e1f6c7{STOPPED}[interval=600000, ownscheduler=false] - STOPPED
|  += HouseKeeper@d554c5f{STOPPED}[interval=600000, ownscheduler=false] - STOPPED
|  += HouseKeeper@7b4c50bc{STOPPED}[interval=660000, ownscheduler=false] - STOPPED
|  += HouseKeeper@1b1cfb87{STOPPED}[interval=660000, ownscheduler=false] - STOPPED
|  += HouseKeeper@28cda624{STOPPED}[interval=660000, ownscheduler=false] - STOPPED
|  += HouseKeeper@7d3e8655{STOPPED}[interval=600000, ownscheduler=false] - STOPPED
|  += HouseKeeper@12591ac8{STARTED}[interval=600000, ownscheduler=true] - STARTED
```

### CODE REVIEW

**Review by sbordet — CHANGES_REQUESTED** [review]
Missing a test case? I really would like to see the "start/stop N times" and verify that we don't leak -- it will be useful for future changes.

File: jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java [hunk 1]
```diff
@@ -123,6 +125,19 @@ public static FilterHolder ensureFilter(ServletContext servletContext)
...
+            {
+                @Override
+                public void lifeCycleStopping(LifeCycle event)
+                {
+                    FilterHolder[] holders = Arrays.stream(servletHandler.getFilters())
+                        .filter(h -> h != holder)
+                        .toArray(FilterHolder[]::new);
+                    servletHandler.setFilters(holders);
+                }
+            });
```
Comment (lachlan-roberts) [inline jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java:137]: @sbordet not sure about this one. Seems like we should be removing the `WebSocketUpgradeFilter` if it was automatically created and not added manually.  But this doesn't seem like the right way to do it.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/DefaultSessionIdManager.java [hunk 2]
```diff
@@ -339,6 +339,7 @@ protected void doStop() throws Exception
         _houseKeeper.stop();
         if (_ownHouseKeeper)
         {
+            removeBean(_houseKeeper);
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/DefaultSessionIdManager.java:342]: Good catch.

File: jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java [hunk 3]
```diff
@@ -90,17 +90,20 @@ public static WebSocketComponents ensureWebSocketComponents(Server server, Servl
...
+        ContextHandler contextHandler = Objects.requireNonNull(ContextHandler.getContextHandler(servletContext));
+        contextHandler.addManaged(serverComponents);
+        contextHandler.addEventListener(new LifeCycle.Listener()
         {
             @Override
-            public void contextDestroyed(ServletContextEvent sce)
+            public void lifeCycleStopping(LifeCycle event)
             {
-                LifeCycle.stop(serverComponents);
+                contextHandler.removeBean(serverComponents);
```
Comment (janbartel) [inline jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java:102]: Do you need to stop the serverComponents?

File: jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java [hunk 4]
```diff
@@ -90,17 +90,20 @@ public static WebSocketComponents ensureWebSocketComponents(Server server, Servl
...
+                contextHandler.removeEventListener(this);
             }
         });
 
+        servletContext.setAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE, serverComponents);
```
Comment (janbartel) [inline jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java:107]: Do you need to remove this attribute when stopping?

File: jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServerContainer.java [hunk 5]
```diff
@@ -117,6 +120,13 @@ public static JettyWebSocketServerContainer ensureContainer(ServletContext servl
         addBean(sessionTracker);
     }
 
+    @Override
+    public void lifeCycleStopping(LifeCycle event)
+    {
+        contextHandler.removeBean(this);
+        contextHandler.removeEventListener(this);
```
Comment (janbartel) [inline jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServerContainer.java]: I was expecting that the stopping event would reference the same objects as the starting - eg what about the container that was added as a managed bean?

File: jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java (hunk 1 above)
Comment (janbartel) [reply jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java:137]: Is there a mapping you need to remove as well?

**Review by janbartel — CHANGES_REQUESTED** [review]
For the HouseKeeper - that looks good, +1. For the websocket pieces I've left more questions rather than things I think you need to fix.

File: jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java:137]: Good point, I will add this in.

File: jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java:102]: Yes, the `removeBean()` call will stop the `serverComponents`.

File: jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java (hunk 4 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java:107]: Its currently being removed automatically, but it wouldn't hurt to also remove this when stopping.
I will add a test case for this as well.

File: jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServerContainer.java (hunk 5 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServerContainer.java]: `this` is the same container that was added as a managed bean.

Comment (lachlan-roberts) [reply jetty-websocket/websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServerContainer.java]: I think its probably clearer to do this removal from a separate `LifeCycle.Listener` added in `ensureContainer`, the same way it is done in `WebSocketServerComponents`. So I will make that change.

File: jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java (hunk 1 above)
Comment (sbordet) [reply jetty-websocket/websocket-servlet/src/main/java/org/eclipse/jetty/websocket/servlet/WebSocketUpgradeFilter.java:137]: @lachlan-roberts I don't think you should remove it.
If the context is stopped, and then restarted, you first check if the filter is already there, so you won't add it twice.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Comment by sbordet** [comment]
@lachlan-roberts also please check why there is a test failure.

**Comment by janbartel** [comment]
@sbordet not sure I agree about not removing the filter ... generally in jetty if the container sets something up during startup, it should clean it up in stop, that way we know things are consistent on every restart.

For the test failure, its an unrelated, transient problem with stopping the infinispan container during the session tests.

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java [hunk 6]
```diff
@@ -1195,6 +1195,34 @@ public void prependFilterMapping(FilterMapping mapping)
...
+        try (AutoLock ignored = lock())
+        {
+            FilterHolder[] holders = Arrays.stream(getFilters())
+                .filter(h -> h != holder)
+                .toArray(FilterHolder[]::new);
+            setFilters(holders);
+        }
+    }
+
+    public void removeFilterMapping(FilterMapping mapping)
```
Comment (lachlan-roberts) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:1212]: @janbartel I am still removing the `WebSocketUpgradeFilter` when stopping. But am now using these new methods to do it.

**Review by sbordet — APPROVED** [review]

**Pull request merged**